README.md

```diff
@@ -1,55 +1,385 @@
+```ocaml
+# test ();;
+- : int * int list = (15, [1; 1; 1; 1; 1; 2; 1; 1; 1; 2; 2; 2; 2; 1; 2])
+```
 
-## Usage
+and as expected it is not working. The interleaving scenario describes
+previously can still happen. Note that, though it is not observable here, this
+is still better then the previous implementation as we are now race-free (see
+[here](#a-note-about-races-in-ocaml) for a quick note about races in OCaml).
 
-`Saturn` can be installed from `opam`: `opam install saturn`. Sample usage of
-`Work_stealing_deque` is illustrated below.
+As a matter of fact, writting a queue with `push` and `pop` functions that are
+both atomic and parallelism-safe is not as easy as it mays
+
+With `Saturn`,
 
 ```ocaml
-module Ws_deque = Work_stealing_deque.M
+let work_saturn id barrier q () =
+  Barrier.await barrier;
+  for i = 0 to 9 do
+    Saturn.Queue.push q id
+  done
 
-let q = Ws_deque.create ()
+let test_saturn () =
+  let barrier = Barrier.create 2 in
+  let q = Saturn.Queue.create () in
+  let d1 = Domain.spawn (work_saturn 1 barrier q) in
+  let d2 = Domain.spawn (work_saturn 2 barrier q) in
+  Domain.join d1;
+  Domain.join d2;
+  let rec pop_all acc =
+    match Saturn.Queue.pop q with
+    | None -> List.rev acc
+    | Some elt -> pop_all (elt :: acc)
+  in
+  let res = pop_all [] in
+  (List.length res, res)
+```
 
-let () = Ws_deque.push q 100
+Let's run it:
 
-let () = assert (Ws_deque.pop q = 100)
+```ocaml
+# test_saturn ();;
+- : int * int list =
+(20, [2; 2; 1; 2; 2; 2; 2; 2; 2; 1; 2; 1; 2; 1; 1; 1; 1; 1; 1; 1])
 ```
 
+### A note about races in OCaml
+
+Because of the great properties of OCaml 5 memory model (see the
+[OCaml manual](https://v2.ocaml.org/manual/parallelism.html#s%3Apar_mm_easy) for
+more details), not a lot can go worst here: at least, not data corruption or
+segmentation fault like it could happen in other languages.
+
+# Usage
+
+This part provides notes about how to use the provided data structures or more
+exactly, what to not do with it. Two main points are discussed:
+
+- some data structures have restrictions on one domain can do
+- the currently provided data structures are non-composable
+
+## Data structures with domain roles
+
+There are several provided data structures that are meant to be used with a
+certain domain configuration. These restrictions enable to do optimizations but
+that breaks safety properties if not respected. Obviously, these restrictions
+are not only written in documentation but also on the name of the data structure
+itself. Single consumer queue means only one domain should pop.
+
+Let's take the example of `Single_prod_single_cons_queue`. As suggested by the
+name, it should be used with only one domain performing `push` (a producer) and
+one domain performing `pop` (a consumer) at the same time. Having two or more
+domains try to perform one of these operations at the same time will break
+safety properties of the `queue` and more likely result in unexpected
```

Comment:
```suggestion
domains try to perform one of these operations at the same time will break
the safety properties of the `queue` and more than likely result in unexpected
```

Reply:
Would it be correct to say: "...more likely result in unexpected
domains try to perform one of these operations simultaneously will break..."

I'm confused.... with what are they trying "one of these operations at the same time?"
If they're trying only one of the operations .... and at the same time (or simultaneously, if you take my suggestion) ... there needs to be a second thing they're trying it with, if that makes sense. 

Perhaps you mean "Having two or more domains simultaneously perform one of these operations..."  
(either `push` or `pop`, right?)

This second way makes it clear that the two or more domains are performing simultaneously rather than performing two operations simultaneously. (At least in my head!) What do you think, @ILeandersson?

Reply:
I read it as two or more domains trying to either push or pop at the same time, not sure I'm correct though

Reply:
It is meant to say that something can go wrong if there are simultaneously one domain doing a `push` and another domain also doing a `push`

(or same with `pop` instead of `push`)

I have changed the text to : 

```suggestion
Let's take the example of `Single_prod_single_cons_queue`. As suggested by the
name, it should be used with only one domain performing `push` (a producer) and
one domain performing `pop` (a consumer) at the same time. Having two or more
domains simultaneously perform `pop` (or `push`) will break the safety properties
of the `queue` and more likely result in unexpected behaviors.L
```

Is it clearer ?